social_navigation_layers/src/human_layer.cpp

```diff
@@ -0,0 +1,223 @@
+// to inifinite, but practically, any changes lower than the cutoff are
+// simply ignored. Calculate that maximum radius of pixels that need to
+// change surrounding an obstacle for a given cutoff and variance.
+//
+// Args:
+//  inscribed_radius: Radius of inscribed cost region (meters)
+//  cutoff: Cost threshold below which changes are not made to costmap
+//  variance: Exponential variance of cost function
+//
+// Returns:
+//  maximum radius of changes to costmap surrounding a human in meters
+double HumanLayer::get_radius(double inscribed_radius, double cutoff, double variance)
+{
+  return inscribed_radius - variance * log(cutoff / inscribed_cost_);
+}
+
+// Calculates the min/max bounding box of costmap changes for all people
+//
+// This is called by the costmap via the updateBounds method in the
+// parent SocialLayer class. Instead of updating all of the cells for
+// the entire costmap, only update a region of interest. This iterates
+// through all of the people detected and finds the minimum bounding box
+// to contain all of the changes for this update cycle.
+//
+// Output:
+//  (min_x, min_y, max_x, max_y) - Pointers to bounding box of changes
+void HumanLayer::updateBoundsFromPeople(double* min_x, double* min_y, double* max_x, double* max_y)
+{
+  double point = get_radius(inscribed_radius_, cutoff_, variance_);
+  std::list<people_msgs::Person>::iterator p_it;
+  for (p_it = transformed_people_.begin(); p_it != transformed_people_.end(); ++p_it)
+  {
+    people_msgs::Person person = *p_it;
+
+    *min_x = std::min(*min_x, person.position.x - point);
+    *min_y = std::min(*min_y, person.position.y - point);
+    *max_x = std::max(*max_x, person.position.x + point);
+    *max_y = std::max(*max_y, person.position.y + point);
+  }
+}
+
+// Updates the costmap cells within the bounded window
+//
+// This method is called by the costmap on every update cycle. It
+// calculates the cost value for each costmap cell surrounding each
+// human reported and applies them to the costmap.
+//
+// Inputs:
+//  master_grid - Reference to 2D costmap
+//  (min_i, min_j, max_i, max_j) - Bounding box of changes to costmap
+void HumanLayer::updateCosts(costmap_2d::Costmap2D& master_grid,
+                             int min_i, int min_j, int max_i, int max_j)
+{
+  if (people_list_.people.size() == 0)
+    return;
+
+  boost::recursive_mutex::scoped_lock lock(lock_);
+  if (!enabled_)
+    return;
+
+  std::list<people_msgs::Person>::iterator p_it;
+  costmap_2d::Costmap2D* costmap = layered_costmap_->getCostmap();
+  double res = costmap->getResolution();
+  double base = get_radius(inscribed_radius_, cutoff_, variance_);
+  // Width of square bounding box for a person in units of costmap cells
+  unsigned int width = std::max(1, static_cast<int>((2.0 * base) / res));
+
+  for (p_it = transformed_people_.begin(); p_it != transformed_people_.end(); ++p_it)
+  {
+    people_msgs::Person person = *p_it;
```

Comment:
here too you can do like in the updateBounds above with the const reference

Reply:
Done.